File: packages/engine/src/temporal/activities.ts
import { logger } from "observability";
import { executorNode } from "../graph/nodes/executor.node";
import { finishNode } from "../graph/nodes/finish.node";
import { mergeNode } from "../graph/nodes/merge.node";
import { plannerNode } from "../graph/nodes/planner.node";
import { triageNode } from "../graph/nodes/triage.node";
import { validatorNode } from "../graph/nodes/validator.node";
import { EpisodicMemory, LongMemory, ShortMemory, TaskMemory } from "../memory";

import type { AgentState } from "../graph/state";

export async function runAgentStateNode(
  nodeName: string,
  state: any,
): Promise<AgentState> {
  logger.info(
    { nodeName, threadId: state.threadId },
    `Temporal activity starting for node: ${nodeName}`,
  );

  let businessId =
    state.businessConfig?.businessId || state.businessId || "ecommerce";

  if (state.threadId && (businessId === "ecommerce" || !businessId)) {
    try {
      const { getDrizzle, threads } = require("db");
      const { eq } = require("drizzle-orm");
      const drizzle = getDrizzle();
      if (drizzle) {
        const threadRows = await drizzle
          .select()
          .from(threads)
          .where(eq(threads.id, state.threadId))
          .limit(1);
        if (threadRows[0]?.businessId) {
          businessId = threadRows[0].businessId;
          state.businessId = businessId;
          if (state.businessConfig) {
            state.businessConfig.businessId = businessId;
          }
        }
      }
    } catch (err) {
      console.warn(
        "[Temporal Activities] Failed to resolve thread businessId:",
        err,
      );
    }
  }

  // Fetch memories if this is the start of execution (or prepare memories on the state object)
  if (!state.longMemoryFacts || !state.episodicEvents) {
    const longMemory = new LongMemory(state.userId, businessId);
    const episodicMemory = new EpisodicMemory(state.userId, businessId);

    const [longFacts, episodicEvents] = await Promise.all([
      longMemory.searchRelevantFacts(state.input),
      episodicMemory.retrieveEvents(state.input),
    ]);

    state.longMemoryFacts = longFacts;
    state.episodicEvents = episodicEvents;
  }

  // Ensure dynamicConfig / businessConfig is initialized
  if (!state.businessConfig || !state.businessConfig.systemPrompt) {
    let defaultLimit = 100;
    if (businessId === "nike") defaultLimit = 150;
    else if (businessId === "adidas") defaultLimit = 120;

    const { getMerchantDisplayName } = require("types");
    const brandName = getMerchantDisplayName(businessId);

    state.businessConfig = {
      businessId,
      systemPrompt: `You are an advanced, professional AI Customer Support Agent representing ${brandName}. Help users resolve order, shipping, and refund queries.`,
      intents: {
        order_status: { description: "Track or check order delivery status." },
        refund: { description: "Process or request refunds." },
        general_query: { description: "General customer questions." },
      },
      tools: [
        "getOrderStatus",
        "processRefund",
        "takeScreenshot",
        "listUserOrders",
      ],
      executionMode: "plan-and-execute",
      confidenceThresholds: { high: 0.85, mid: 0.6 },
      refundAutoApprovalLimit: defaultLimit,
      ...(state.businessConfig || {}),
    };
  }

  let resultState = { ...state };

  switch (nodeName) {
    case "triage": {
      const updates = await triageNode(state);
      resultState = { ...resultState, ...updates };
      break;
    }
    case "planner": {
      const updates = await plannerNode(state);
      resultState = { ...resultState, ...updates };
      break;
    }
    case "merge": {
      const updates = await mergeNode(state);
      resultState = { ...resultState, ...updates };
      break;
    }
    case "executor": {
      const updates = await executorNode(state);
      resultState = { ...resultState, ...updates };
      break;
    }
    case "validator": {
      const updates = await validatorNode(state);
      resultState = { ...resultState, ...updates };
      break;
    }
    case "finish": {
      const updates = await finishNode(state);
      resultState = { ...resultState, ...updates };

      // Commit memory updates and physically write chat records to Postgres messages table via ShortMemory!
      if (resultState.output) {
        const shortMemory = new ShortMemory(state.threadId);
        const episodicMemory = new EpisodicMemory(state.userId, businessId);
        const longMemory = new LongMemory(state.userId, businessId);

        await shortMemory.addMessage("user", state.input);
        await shortMemory.addMessage("assistant", resultState.output);

        await Promise.all([
          episodicMemory.addEvent(
            `Handled conversation thread: ${state.threadId}. Output summary: ${resultState.output.substring(0, 80)}`,
            5,
          ),
          longMemory.extractAndStoreFact(resultState.output, state.input),
        ]);
      }

      if (resultState.taskPlan) {
        const taskMemory = new TaskMemory(state.threadId);
        await taskMemory.saveTaskState(resultState.taskPlan);
      }
      break;
    }
    default:
      throw new Error(`Unknown node: ${nodeName}`);
  }

  logger.info(
    { nodeName, threadId: state.threadId },
    `Temporal activity finished for node: ${nodeName}`,
  );
  return resultState;
}

File: services/engine-py/src/engine_py/temporal/activities.py
"""Temporal Activities — 镜像 temporal/activities.ts(runAgentStateNode 节点分发)。"""

from __future__ import annotations

from sqlalchemy import select

from ..cards import CardSynthesizer
from ..db import Thread, get_session
from ..graph.nodes import executor_node, finish_node, merge_node, planner_node, triage_node, validator_node
from ..graph.state import AgentState, to_ts_dict
from ..memory import EpisodicMemory, LongMemory, ShortMemory, TaskMemory
from ..tenant import get_merchant_display_name

_NODE_HANDLERS = {
    "triage": triage_node,
    "planner": planner_node,
    "merge": merge_node,
    "executor": executor_node,
    "validator": validator_node,
    "finish": finish_node,
}


async def run_agent_state_node(node_name: str, ts_state: dict) -> dict:
    """以 TS camelCase 状态在指定节点上推进一步(进出均为 TS 状态格式)。"""
    state: AgentState = dict(ts_state)  # type: ignore[assignment]
    business_id = str(
        (state.get("businessConfig") or {}).get("businessId") or state.get("businessId") or "ecommerce"
    )

    if state.get("threadId") and (business_id == "ecommerce" or not business_id):
        try:
            async with get_session() as session:
                row = (
                    await session.execute(select(Thread).where(Thread.id == state["threadId"]).limit(1))
                ).scalar_one_or_none()
                if row and row.business_id:
                    business_id = row.business_id
        except Exception as err:
            print(f"[Temporal Activities] Failed to resolve thread businessId: {err}")

    if not state.get("longMemoryFacts") or not state.get("episodicEvents"):
        long_memory = LongMemory(state.get("userId", ""), business_id)
        episodic_memory = EpisodicMemory(state.get("userId", ""), business_id)
        long_facts, episodic_events = await __import__("asyncio").gather(
            long_memory.search_relevant_facts(state.get("input", "")),
            episodic_memory.retrieve_events(state.get("input", "")),
        )
        state["longMemoryFacts"] = long_facts
        state["episodicEvents"] = episodic_events

    if not state.get("businessConfig") or not (state.get("businessConfig") or {}).get("systemPrompt"):
        default_limit = 150 if business_id == "nike" else 120 if business_id == "adidas" else 100
        brand_name = get_merchant_display_name(business_id)
        state["businessConfig"] = {
            "businessId": business_id,
            "systemPrompt": (
                f"You are an advanced, professional AI Customer Support Agent representing {brand_name}. "
                "Help users resolve order, shipping, and refund queries."
            ),
            "intents": {
                "order_status": {"description": "Track or check order delivery status."},
                "refund": {"description": "Process or request refunds."},
                "general_query": {"description": "General customer questions."},
            },
            "tools": ["getOrderStatus", "processRefund", "takeScreenshot", "listUserOrders"],
            "executionMode": "plan-and-execute",
            "confidenceThresholds": {"high": 0.85, "mid": 0.6},
            "refundAutoApprovalLimit": default_limit,
            **(state.get("businessConfig") or {}),
        }

    # 将 TS 状态转 Python 状态键推进节点
    from ..graph.state import AgentState as _AS  # noqa: F401 — 类型提示

    py_state: dict = _ts_to_py_state(state)
    handler = _NODE_HANDLERS.get(node_name)
    if handler is None:
        raise ValueError(f"Unknown node: {node_name}")
    updates = await handler(py_state)  # type: ignore[arg-type]
    py_state.update({k: v for k, v in updates.items()})

    if node_name == "finish":
        synthesized_cards = CardSynthesizer.synthesize_cards(
            {
                "taskPlan": py_state.get("task_plan"),
                "intents": py_state.get("intents"),
                "damageAssessment": py_state.get("damage_assessment"),
            }
        )
        final_cards = py_state.get("cards") or synthesized_cards
        py_state["cards"] = final_cards

        if py_state.get("output"):
            short_memory = ShortMemory(state.get("threadId", ""))
            episodic_memory = EpisodicMemory(state.get("userId", ""), business_id)
            long_memory = LongMemory(state.get("userId", ""), business_id)
            await short_memory.add_message("user", state.get("input", ""))
            await short_memory.add_message("assistant", py_state["output"], final_cards)
            await __import__("asyncio").gather(
                episodic_memory.add_event(
                    f"Handled conversation thread: {state.get('threadId')}. "
                    f"Output summary: {py_state['output'][:80]}",
                    5,
                ),
                long_memory.extract_and_store_fact(py_state["output"], state.get("input", "")),
            )
        if py_state.get("task_plan"):
            await TaskMemory(state.get("threadId", "")).save_task_state(py_state["task_plan"])

    return _py_to_ts_state(py_state)


_TS_TO_PY = {
    "threadId": "thread_id",
    "userId": "user_id",
    "jobId": "job_id",
    "imageUrls": "image_urls",
    "damageAssessment": "damage_assessment",
    "inputEmbedding": "input_embedding",
    "activeDomainRole": "active_domain_role",
    "guideContext": "guide_context",
    "cartContext": "cart_context",
    "orderContext": "order_context",
    "taskPlan": "task_plan",
    "shortMemory": "short_memory",
    "longMemoryFacts": "long_memory_facts",
    "episodicEvents": "episodic_events",
    "ragDocuments": "rag_documents",
    "businessConfig": "business_config",
    "loopCount": "loop_count",
    "globalTransitionsCount": "global_transitions_count",
    "toolErrorsCount": "tool_errors_count",
}


def _ts_to_py_state(ts_state: dict) -> dict:
    return {_TS_TO_PY.get(k, k): v for k, v in ts_state.items()}


def _py_to_ts_state(py_state: dict) -> dict:
    return to_ts_dict(py_state)
